### PR TITLE
Feature grafana

### PR DESCRIPTION
**Pull request by maaikelimper**
The "monitoring"-branch featuring:
- a separate .yml for the monitoring-stack: "docker-compose.monitoring.yml" (which needs to be included when starting the wis2box to enable monitoring, see change in wis2box-ctrl.py)
- 2 containers to publish metrics: metrics_collector and mqtt_metrics_collector (exposed on port 8000 and 8001)
- prometheus-container
- loki-container
- grafana-container (with "home"-dashboard to show wis2box workflow monitoring) exposed on port 3000
- overrides the logging of existing wis2box-containers to redirect logs to loki
- some very basic documentation

### CODE REVIEW

File: docker/metrics_collector/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,15 @@
+FROM python:3-alpine
```
Comment (tomkralidis) [inline docker/metrics_collector/Dockerfile:1]: Add license header (example: https://github.com/wmo-im/wis2box/blob/main/docker/data-consumer/Dockerfile#L1-L20)

File: docker/metrics_collector/metrics_collector.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
+#!/usr/bin/env python
```
Comment (tomkralidis) [inline docker/metrics_collector/metrics_collector.py:1]: Add license header (example: https://github.com/wmo-im/wis2box/blob/main/docker/data-consumer/Dockerfile#L1-L20)

File: docker/metrics_collector/metrics_collector.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+logging.basicConfig(stream=sys.stdout)
+# create our own logger using same log-level as wis2box
+logger = logging.getLogger('metrics_collector')
+logger.setLevel(WIS2BOX_LOGGING_LOGLEVEL)
+
+INTERRUPT = False
+
+def collect_file_data_by_topic(root,filematch):
+    """
+    helper-function to collect data using labels 'topic' from given root-path and file-type
```
Comment (tomkralidis) [inline docker/metrics_collector/metrics_collector.py:30]: Add function annotation and docstring (example: https://github.com/wmo-im/wis2box/blob/main/wis2box/data/__init__.py#L67-L74)

File: docker/metrics_collector/metrics_collector.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    bufr_public_bytes_gauge = Gauge('bufr4_public_bytes', 'Bytes used by bufr4 stored in /data/public')
+    bufr_public_bytes_by_topic_gauge = Gauge('bufr4_public_bytes_by_topic', 'Bytes used by bufr4 stored in /data/public, by topic',["topic"])
+
+    csv_incoming_nrfile_gauge = Gauge('csv_incoming_nrfile', 'Number of csv in /data/incoming')
+    csv_incoming_nrfile_by_topic_gauge = Gauge('csv_incoming_nrfile_by_topic', 'Number of csv in /data/incoming, by topic', ["topic"])
+    csv_incoming_bytes_gauge = Gauge('csv_incoming_bytes', 'Bytes used by csv stored in /data/incoming')
+    csv_incoming_bytes_by_topic_gauge = Gauge('csv_incoming_bytes_by_topic', 'Bytes used by csv stored in /data/incoming, by topic', ["topic"])
+    
+    
+    while INTERRUPT == False:
```
Comment (tomkralidis) [inline docker/metrics_collector/metrics_collector.py:63]: `while not INTERRUPT:`

File: docker/metrics_collector/metrics_collector.py (hunk 2 above)
Comment (tomkralidis) [inline docker/metrics_collector/metrics_collector.py:1]: Remove shebang line

File: docker/mqtt_metrics_collector/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,15 @@
+FROM python:3-alpine
```
Comment (tomkralidis) [inline docker/mqtt_metrics_collector/Dockerfile:1]: Add license header (example: https://github.com/wmo-im/wis2box/blob/main/docker/data-consumer/Dockerfile#L1-L20)

File: docker/mqtt_metrics_collector/mqtt_metrics_collector.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
```
Comment (tomkralidis) [inline docker/mqtt_metrics_collector/mqtt_metrics_collector.py:1]: Add license header (example: https://github.com/wmo-im/wis2box/blob/main/docker/data-consumer/Dockerfile#L1-L20)

Comment (tomkralidis) [inline docker/mqtt_metrics_collector/mqtt_metrics_collector.py:1]: Remove shebang line

File: docker/metrics_collector/metrics_collector.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                    data[topic]['size'] += size
+                    data[topic]['file_count'] += 1
+                else :  
+                    data[topic] = {}
+                    data[topic]['size'] = size
+                    data[topic]['file_count'] = 1
+    return data
+
+def gather_file_metrics():
+    """
```
Comment (tomkralidis) [inline docker/metrics_collector/metrics_collector.py:49]: Add function annotation and docstring (example: https://github.com/wmo-im/wis2box/blob/main/wis2box/data/__init__.py#L67-L74)

File: docker/mqtt_metrics_collector/mqtt_metrics_collector.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+WIS2BOX_LOGGING_LOGLEVEL = os.environ.get('WIS2BOX_LOGGING_LOGLEVEL')
+# gotta log to stdout so docker logs sees the python-logs
+logging.basicConfig(stream=sys.stdout)
+# create our own logger using same log-level as wis2box
+logger = logging.getLogger('mqtt_metrics_collector')
+logger.setLevel(WIS2BOX_LOGGING_LOGLEVEL)
+
+INTERRUPT = False
+
+def sub_connect(client, userdata, flags, rc, properties=None):
```
Comment (tomkralidis) [inline docker/mqtt_metrics_collector/mqtt_metrics_collector.py:36]: Add function annotation and docstring (example: https://github.com/wmo-im/wis2box/blob/main/wis2box/data/__init__.py#L67-L74)

File: docker/mqtt_metrics_collector/mqtt_metrics_collector.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    print("on connection to subscribe:", mqtt.connack_string(rc))
```
Comment (tomkralidis) [inline docker/mqtt_metrics_collector/mqtt_metrics_collector.py:37]: Should this be printing, or emitting a logging statement?

File: docker/mqtt_metrics_collector/mqtt_metrics_collector.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    for s in ["xpublic/#"]:
+        client.subscribe(s, qos=1)
+
+mqtt_msg_counter = Counter('mqtt_msg_count', 'Number of messages seen on MQTT')
+mqtt_msg_by_topic_counter = Counter('mqtt_msg_count_by_topic', 'Number of messages seen on MQTT, by topic', ["topic"])
+
+def sub_mqtt_metrics(client, userdata, msg):
```
Comment (tomkralidis) [inline docker/mqtt_metrics_collector/mqtt_metrics_collector.py:44]: Add function annotation and docstring (example: https://github.com/wmo-im/wis2box/blob/main/wis2box/data/__init__.py#L67-L74)

File: docker/mqtt_metrics_collector/mqtt_metrics_collector.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    """
+      subscribe and update MQTT metrics for each new message received
+    """
+    m = json.loads(msg.payload.decode('utf-8'))
+    logger.info(f"Received message on topic ={msg.topic}")
+    mqtt_msg_by_topic_counter.labels(msg.topic).inc(1)
+    mqtt_msg_counter.inc(1)
+
+def gather_mqtt_metrics():
```
Comment (tomkralidis) [inline docker/mqtt_metrics_collector/mqtt_metrics_collector.py:53]: Add function annotation and docstring (example: https://github.com/wmo-im/wis2box/blob/main/wis2box/data/__init__.py#L67-L74)

File: docker/mqtt_metrics_collector/mqtt_metrics_collector.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    # explicitly set the counter to 0 at the start
+    mqtt_msg_counter.inc(0)
+
+    # parse username and password out of the WIS2BOX_BROKER variable
+    BROKER = os.environ.get('WIS2BOX_BROKER')
+    
+    # generate a random clientId for the mqtt-session
+    r = random.Random()
+    client_id = f"mqtt_metrics_collector_{r.randint(1,1000):04d}"
+    logger.info(f"{BROKER}")
```
Comment (tomkralidis) [inline docker/mqtt_metrics_collector/mqtt_metrics_collector.py:66]: `logger.info(BROKER)`

File: docs/source/monitoring/index.rst [hunk 13]
```diff
@@ -0,0 +1,32 @@
+.. _monitoring:
+
+Monitoring
+==========
+
+wis2box has built-in monitoring functions based on prometheus, loki and grafana. The Grafana-endpoint is exposed on the localhost at port 3000.
```
Comment (tomkralidis) [inline docs/source/monitoring/index.rst:6]: Add links to the homepages of these tools (prometheus, loki and grafana)

**Review by tomkralidis — CHANGES_REQUESTED** [review]
Excellent work @maaikelimper!  Please see minor change requests.  Thanks

File: docker/mqtt_metrics_collector/mqtt_metrics_collector.py (hunk 9 above)
Comment (maaikelimper) [reply docker/mqtt_metrics_collector/mqtt_metrics_collector.py:37]: it should log to info, changed it

**Review by tomkralidis — APPROVED** [review]

**Pull request merged**